### ISSUE TITLE
Fails to pass array variable as a function argument.

### ISSUE THREAD

**Issue by purna135** [lcompilers/lpython#351]
Here is an example with an error message.
```python
n: i32
n = TypeVar('n')

def foo(n:i32, arr: i32[n]):
    pass


A: i32[5]
A = empty(5)
foo(5, A)
```

Error:
```python
(LPython) purna@hotdev:~/OpenSource/lpython$ ; ModuleID = 'LFortran'
source_filename = "LFortran"

%array = type { i32*, i32, %dimension_descriptor*, i1, i32 }
%dimension_descriptor = type { i32, i32, i32, i32 }

@A = global i32 0
@n = global i32 0

define void @_lpython_main_program() {
.entry:
  %call_arg_value = alloca i32, align 4
  store i32 5, i32* %call_arg_value, align 4
  call void @foo(i32* %call_arg_value, i32* @A)
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define void @foo(i32* %n, %array* %arr) {
.entry:
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define i32 @main() {
.entry:
  call void @_lpython_main_program()
  ret i32 0
}
code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
i32* @A
 %array*  call void @foo(i32* %call_arg_value, i32* @A)



Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).

```

**Comment by certik** [comment]
Thanks for the report! I think that this should work, so we need to fix it.

**Comment by purna135** [comment]
I am trying to fix it. I'm referring to your video of [live coding](https://www.youtube.com/watch?v=yuYsyM08bss). It would be extremely helpful if you could point me in the right direction.

**Comment by certik** [comment]
Yes, the video is the right direction. Just follow the steps there to figure out where in LPython the problem is. Is it in ASR or in LLVM? Then let's see what the correct ASR should be for this, assuming the problem is in AST to ASR translation.

**Comment by purna135** [comment]
Sure, Thank you : )

**Comment by purna135** [comment]
Hello, @certik. I watched the entire video but am still unable to determine where I need to change based on the error message.
The issue is in ASR, so I looked at asr to llvm.cpp but couldn't figure out which code block to change. Please assist me.

**Comment by certik** [comment]
This might be one of the harder things to fix as the first patch. I looked into it, the ASR seems OK, but there seems to be a bug in the asr to llvm (so you were looking at the right file). In the generated LLVM, the definition of `foo` is:
```
define void @foo(i32* %n, %array* %arr)
```
That seems ok. The call is:
```
  call void @foo(i32* %call_arg_value, i32* @A)
```
Which is incompatible in the second argument, as the LLVM verify pass says also.

This is a bug how our LLVM backend handles global variables. In the meantime a workaround is to use a local function:
```
n: i32
n = TypeVar('n')

def foo(n:i32, arr: i32[n]):
    pass


def main():
    A: i32[5]
    A = empty(5)
    foo(5, A)

main()
```
This works.

CC @czgdp1807. I think I've seen this bug before, probably not difficult to fix.

**Comment by czgdp1807** [comment]
Well, we need to see two things here. What's the difference between the ASRs of code in the OP and the alternative provided by @certik in the above comment. Then we can figure out if a fix is needed in AST to ASR transition or in ASR to LLVM transition. Most probably if the alternative works then I think there is a bug in AST to ASR transition. If not then we can see what's going in ASR to LLVM by comparing LLVM outputs of both the codes and figuring out where things go wrong.

**Comment by purna135** [comment]
Thank you so much, the above code is working fine : )
